### PR TITLE
Configurable Clusterpool Initial Machinesets

### PR DESCRIPTION
**Pull request by hassenius**
This is a proposal to allow customizing the controlplane and worker machines in a clusterpools `install-config.yaml`, addressing #1178 

It allows this through defining an optional `spec.controlPlaneMachinePool` and/or `spec.workerMachinePool` which will override the default values in the 

Example:
```yaml
kind: ClusterPool
metadata:
  name: hktest
  namespace: hive
spec:
  controlPlaneMachinePool:
    aws:
      type: m5.xlarge
      rootVolume:
        size: 128
        iops: 200
        type: gp2
  workerMachinePool:
    aws:
      type: m5.xlarge
      rootVolume:
        size: 128
        iops: 200
        type: gp2
  platform:
    aws:
      credentialsSecretRef:
        name: global-aws-creds
      region: eu-west-1
```

Initially support only added to AWS, Azure and GCP.

Hacked together with little hands-on experience with Go. Expecting things to be missing and sub-optimal, but it does compile and produce the desired result.

### CODE REVIEW

File: pkg/clusterresource/aws.go [hunk 1]
```diff
@@ -107,9 +107,19 @@ func (p *AWSCloudBuilder) addInstallConfigPlatform(o *Builder, ic *installertype
...
+		ic.ControlPlane.Platform.AWS = o.ControlPlaneMachinePool.AWS
+	} else {
+		ic.ControlPlane.Platform.AWS = mpp
+	}
+
+	if (o.WorkerMachinePool != installertypes.MachinePoolPlatform{}) {
+		ic.Compute[0].Platform.AWS = o.WorkerMachinePool.AWS
+	} else {
+		ic.Compute[0].Platform.AWS = mpp
+	}
```
Comment (dgoodwin) [inline pkg/clusterresource/aws.go:122]: This covers the installconfig but I think you're then going to conflict with the Hive MachinePool that takes over day 2 operation, see addMachinePoolPlatform above. You'll want to update this to use the same settings. Per comments above the API should use the Hive MachinePoolPlatform anyhow so that one will be easy, but this part for the install config will be where you need to explicitly translate what we support.

**Review by dgoodwin — COMMENTED** [review]
This is great for a first stab at go programming. Nice work.

I will check with the team to see if there are any fundamental concerns about exposing this on ClusterPool.

File: pkg/apis/hive/v1/clusterpool_types.go [hunk 2]
```diff
@@ -33,6 +34,14 @@ type ClusterPoolSpec struct {
 	// claimed will not be affected when this value is modified.
 	// +optional
 	Labels map[string]string `json:"labels,omitempty"`
+
+	// WorkerMachinePool allows defining a machinepool declaration for new worker nodes
+	// +optional
+	WorkerMachinePool installertypes.MachinePoolPlatform `json:"workerMachinePool,omitempty"`
```
[the remark this answers is not in the document]
Comment (hassenius) [reply pkg/apis/hive/v1/clusterpool_types.go]: I've been thinking about this for a while. I _think_ we're basically passing it straight into `*installertypes.InstallConfig`:  [aws.go#L13](https://github.com/hassenius/hive/blob/clusterpool-settings/pkg/clusterresource/aws.go#L13)  [aws.go#L93](https://github.com/hassenius/hive/blob/clusterpool-settings/pkg/clusterresource/aws.go#L93)

Would I be right in guessing that any potential upstream changes to `installertypes.MachinePoolPlatform` would be kept in sync with `installertypes.InstallConfig`?

Would we still need to put something in between?

[the remark this answers is not in the document]
Comment (dgoodwin) [reply pkg/apis/hive/v1/clusterpool_types.go]: Yes ultimately that is what we do but we don't want a vendor update to inadvertently change our API we expose, potentially showing a feature that is not available on older versions of OpenShift that we still need to be able to install via Hive. For this reason we don't publically expose installer types in our API rather we have to wrap them and translate. New changes to their API may or may not get mirrored in ours, depending on whether or not we need it.

**Comment by hassenius** [comment]
Thanks for your consideration and review @dgoodwin .

I've implemented the _easy stuff_, and now I'm trying to think through the implications of the hivev1.MachinePoolPlatform to the best of my ability.

If I'm on the right track with this I'd also like to look at exposing the number of workers.

**Comment by dgoodwin** [comment]
There is a worker count exposed on the resource builder (WorkerNodeCounts) but not on the ClusterPool.

Now full disclosure, some concerns on the team if we can support this, we don't want to redo the InstallConfig, so we need to figure out how to handle what could be a lot more customization. An alternative might be specifying an install config secret and we splice in the things that change per cluster. (clustername for example)

**Comment by hassenius** [comment]
Appreciate that. I'll hold back till I hear the outcome.

**Comment by hassenius** [comment]
Wonder if you have had a chance to discuss this @dgoodwin ?

We really need the ability to have configurable cluster pools for our use case, and would really prefer not having to maintain a separate fork if it's at all avoidable..

**Comment by dgoodwin** [comment]
I'm sorry I forgot all about it. I will raise with the team this afternoon.

**Comment by hassenius** [comment]
Any updates on this?

It would be useful to know the current thinking, such that if we do need to maintain a separate fork that it's as closely aligned with upstream as possible to lessen the maintenance burden and make for an easier transition if/when the capability is added upstream.

**Comment by dgoodwin** [comment]
Terribly sorry I can't seem to keep this in my head long enough to bring it up one of our calls.

The fact that this covers both control plane and worker is where I'm struggling to know what to do here. We have a Hive MachinePool API in play for worker nodes, but not for the control plane. Your desired settings must span both the install-config, and should also affect the generated Hive MachinePool. 

The change you have here exposes control plane settings that Hive literally doesn't expose on a normal ClusterDeployment, they're only in the install-config secret, which seems like a bad idea. But your worker changes here expose the normal things for a Hive MachinePool, which is good.

I fear this means we need both, drop the control plane config in this PR but keep the worker config, and add in some kind of install-config Secret templating, which is really unfortunate as that's not an area I want to us to get into as we have to worry about different versions of the install-config if users are providing them and expecting us to modify them on the fly. Right now we just generate one version, and that should always work, but if the installer changes and has new versions we don't know about or have vendored, we'll have problems, and hive is supposed to work as well as possible even with future versions of the installer as we don't want the release cadence coupled. We could perhaps reject all but the one version of install-config we work with.

Or we could just be more surgical. You can specify an install config Secret, if you do, we will surgically edit specific yaml/json properties to match the settings on the ClusterPool. Today I think that list is baseDomain, platform, pullSecret, and probably the WorkerPlatform this PR adds.

TL;DR I'm thinking keep the ClusterPool.Spec.WorkerPlatform this PR adds, drop the ControlPlane, and introduce ClusterPool.Spec.InstallConfigTemplateRef to point to a Secret with a fully formatted InstallConfig, which we will use instead of generating our own and surgically splice in the fields you configured on your ClusterPool.

How does this sound? 

cc @openshift/openshift-team-hive @abhinavdahiya

**Comment by hassenius** [comment]
Thanks for getting back to me @dgoodwin, I think this sounds like a good approach that would fit our use cases. I'll take the ControlPlane out of this PR and make sure it's aligned with current implementations. 

Then I'll look at the installconfig and open a separate PR for that.

**Comment by hassenius** [comment]
Took me a while to get to this, but I reversed the order to my original plan and opened this: #1214 to address the suggestion of surgical modification of a `install-config.yaml`, and then we can re-assess the need for `ClusterPool.Spec.WorkerPlatform` after that....

**Comment by hassenius** [comment]
@dgoodwin wonder are you still interested in this one? Seeing that https://github.com/openshift/hive/pull/1214 is merged, my teams needs are already met -- but seeing that the bulk of the work on this PR has already been done, I would be happy to spend a little bit of time to do a rebase + make the suggested adjustments if the work can benefit someone...

**Comment by dgoodwin** [comment]
We don't have an immediate need but I suspect it would benefit someone at some point, and the change seems logical and useful to me. If you want to do the work I think we'll take the PR as no one has raised objections yet.

**Comment by hassenius** [comment]
Don't have spare cycles to work on this, so will close this
